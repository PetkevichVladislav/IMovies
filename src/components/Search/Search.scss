.search{
    display: flex;
    justify-content: start;
    align-items: center;
    align-self: flex-start;
    width: 100%;
    
    &__input{
        background-color: rgba(50, 50, 50, 0.8);;
        color: white;
        border: none;
        padding-left: 0.5rem;
        box-sizing: border-box;
        border-radius: 5px;
        margin-right: 0.5rem;
        font-size: 20px;
        width: 100%;
        height: 57px;
    }

    &__input::placeholder{
        color: #555555;
        font-size: 20px;
    }

    &__input:focus{
        outline: none;
    }

    &__button{
        background-color: #F65261;
        color: white;
        border: none;
        border-radius: 5px;
        width: 233px;
        height: 57px;
        font-size: 20px;
    }
}
