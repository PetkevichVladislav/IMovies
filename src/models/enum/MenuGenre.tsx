export enum Genre{
    All = "ALL",
    Documentary = "DOCUMENTARY",
    Comedy = "COMEDY",
    Horror = "HORROR",
    Crime = "CRIME",
}