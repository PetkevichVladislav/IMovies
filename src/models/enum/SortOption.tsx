export enum SortOption {
    ReleaseDate = "RELEASE DATE",
    Title = "TITLE"
}
