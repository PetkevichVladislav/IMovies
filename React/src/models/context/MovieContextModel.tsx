import { MovieModel } from "../MovieModel";
import { Genre } from "../enum/MenuGenre";
import { SortOption } from "../enum/SortOption";

export interface MovieContextModel{
    selectedGenre: string,
    setSelectedGenre: (genre: Genre) => void,
    moviesQuantity: number,
    movieList: MovieModel[],
    searchQuery: string,
    setSearchQuery: (query: string) => void,
    sortCriterion: SortOption,
    setSortCriterion: (sortCriterion : SortOption) => void,
    selectedMovieId?: string | null | undefined,
    setSelectedMovieId?: React.Dispatch<React.SetStateAction<string | null | undefined>>,
}