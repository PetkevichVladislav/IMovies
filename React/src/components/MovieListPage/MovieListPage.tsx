import MainPart from '../Main/Main';
import Footer from '../Footer/Footer';

import { Genre } from "../../models/enum/MenuGenre";
import { MovieModel } from "../../models/MovieModel";
import { SortOrder } from "../../models/enum/SortOrder";
import { SortOption } from "../../models/enum/SortOption";
import { getMovies } from "../../services/MovieRequestService";
import { SearchOptions } from "../../models/enum/SearchOptions";
import { MovieContextModel } from "../../models/context/MovieContextModel";
import { GetMoviesRequestModel } from "../../models/request/GetMoviesRequestModel";
import { mapSortOptionToRequestSortOption } from "../../services/SortOptionMapService";
import { FC, ReactNode, createContext, useCallback, useContext, useEffect, useRef, useState } from "react";


export const MovieContext = createContext<MovieContextModel | null>(null);

export const useMovieContext = (): MovieContextModel | null => useContext(MovieContext);

function capitalizeWords(string: string): string {
    return string
        .split(" ")
        .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
        .join(" ");
}

function stringifyParametersToUrlQuery(parameters: SearchQueryParameters): string {
    return Object.keys(parameters).map(function (key) {
        const value = parameters[key as keyof SearchQueryParameters]?.toLocaleLowerCase();
        if (value !== "") {
            return key + '=' + parameters[key as keyof SearchQueryParameters]?.toLocaleLowerCase();
        }
        else {
            return "";
        }
    }).join('&');
}

interface SearchQueryParameters {
    query?: string | null,
    sorting?: SortOption | null,
    genre?: Genre | null,
}

interface MovieListProps {
    children: ReactNode,
    searchParams: {
        query?: string | null,
        sorting?: SortOption | null,
        genre?: Genre | null,
    }
}

export const MovieListPage: FC<MovieListProps> = ({ children, searchParams }) => {
    const [query, setQuery] = useState<string>(searchParams.genre ?? "");
    const [genre, setGenre] = useState<Genre>(searchParams.genre?.toLocaleUpperCase() as Genre ?? Genre.All);
    const [sorting, setSorting] = useState<SortOption>(searchParams.sorting as SortOption ?? SortOption.ReleaseDate);

    const [movieList, setMovieList] = useState<MovieModel[]>([]);
    const [moviesQuantity, setMovieQuantity] = useState<number>(0);
    const firstRender = useRef(true);

    const filteringAndSortingAbortController = useRef<AbortController | null>(null);

    // First rendering.
    useEffect(() => {
        const movieRequestParameters: GetMoviesRequestModel = {
            sortBy: mapSortOptionToRequestSortOption(sorting),
            sortOrder: SortOrder.Descending,
        };

        getMovies(movieRequestParameters).then(movieModels => {
            setMovieList(movieModels);
            setMovieQuantity(movieModels.length)
        });
    }, []);

    // Update sorting and ordering options.
    useEffect(() => {
        if (firstRender.current) {
            firstRender.current = false;
            return;
        }

        if (filteringAndSortingAbortController.current) {
            filteringAndSortingAbortController.current.abort();
        }

        const requestModel: GetMoviesRequestModel = {
            sortBy: mapSortOptionToRequestSortOption(sorting),
            sortOrder: SortOrder.Descending,
            search: capitalizeWords(genre),
            searchBy: SearchOptions.Genres,
        };

        if (query || genre === Genre.All) {
            requestModel.search = query;
            requestModel.searchBy = SearchOptions.Title;
        }

        filteringAndSortingAbortController.current = new AbortController();
        const requestParameters: RequestInit = {
            signal: filteringAndSortingAbortController.current!.signal,
        }

        getMovies(requestModel, requestParameters).then().then(movieModels => {
            setMovieList(movieModels);
            setMovieQuantity(movieModels.length)
        })
    }, [sorting, query, genre, searchParams, firstRender, filteringAndSortingAbortController]);

    const movieContextModel: MovieContextModel = {
        selectedGenre: genre,
        setSelectedGenre: (genre: Genre) => {
            setQuery("");
            setGenre(genre);
        },
        moviesQuantity: moviesQuantity,
        movieList: movieList,
        searchQuery: query,
        setSearchQuery: (query: string) => {
            setGenre(Genre.All);
            setQuery(query);
        },
        sortCriterion: sorting,
        setSortCriterion: (sorting: SortOption) => {
            setSorting(sorting);
        },
    }

    return (
        <>
            <MovieContext.Provider value={movieContextModel}>
                {children}
                <hr className="app__divider" />
                <MainPart />
                <Footer />
            </MovieContext.Provider>
        </>
    );
}
