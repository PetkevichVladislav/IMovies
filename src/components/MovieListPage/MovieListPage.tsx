import MainPart from '../Main/Main';
import Footer from '../Footer/Footer';

import { Genre } from "../../models/enum/MenuGenre";
import { MovieModel } from "../../models/MovieModel";
import { SortOrder } from "../../models/enum/SortOrder";
import { SortOption } from "../../models/enum/SortOption";
import { getMovies } from "../../services/MovieRequestService";
import { SearchOptions } from "../../models/enum/SearchOptions";
import { MovieContextModel } from "../../models/context/MovieContextModel";
import { GetMoviesRequestModel } from "../../models/request/GetMoviesRequestModel";
import { mapSortOptionToRequestSortOption } from "../../services/SortOptionMapService";
import { FC, createContext, useCallback, useContext, useEffect, useRef, useState } from "react";
import { Outlet, useSearchParams } from "react-router-dom";


export const MovieContext = createContext<MovieContextModel | null>(null);

export const useMovieContext = (): MovieContextModel | null => useContext(MovieContext);

function capitalizeWords(string: string): string {
    return string
        .split(" ")
        .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
        .join(" ");
}

function stringifyParametersToUrlQuery(parameters: SearchQueryParameters): string {
    return Object.keys(parameters).map(function (key) {
        const value = parameters[key as keyof SearchQueryParameters]?.toLocaleLowerCase();
        if (value !== "") {
            return key + '=' + parameters[key as keyof SearchQueryParameters]?.toLocaleLowerCase();
        }
        else {
            return "";
        }
    }).join('&');
}

interface SearchQueryParameters {
    query?: string | null,
    sorting?: SortOption | null,
    genre?: Genre | null,
}

export const MovieListPage: FC = () => {
    const [searchParams, setSearchQueryParams] = useSearchParams();
    const query = searchParams.get("query") ?? "";
    const sorting = searchParams.get("sorting")?.toLocaleUpperCase() as SortOption ?? SortOption.ReleaseDate;
    const genre = searchParams.get("genre")?.toLocaleUpperCase() as Genre ?? Genre.All;
    const setSearchParams = useCallback((parameters: SearchQueryParameters) => {
        const queryString = stringifyParametersToUrlQuery(parameters);
        setSearchQueryParams(queryString);
    }, [setSearchQueryParams]);

    const [movieList, setMovieList] = useState<MovieModel[]>([]);
    const [moviesQuantity, setMovieQuantity] = useState<number>(0);
    const firstRender = useRef(true);

    const filteringAndSortingAbortController = useRef<AbortController | null>(null);

    // First rendering.
    useEffect(() => {
        const movieRequestParameters: GetMoviesRequestModel = {
            sortBy: mapSortOptionToRequestSortOption(sorting),
            sortOrder: SortOrder.Descending,
        };

        getMovies(movieRequestParameters).then(movieModels => {
            setMovieList(movieModels);
            setMovieQuantity(movieModels.length)
        });
    }, []);

    // Update sorting and ordering options.
    useEffect(() => {
        if (firstRender.current) {
            firstRender.current = false;
            return;
        }

        if (filteringAndSortingAbortController.current) {
            filteringAndSortingAbortController.current.abort();
        }

        const requestModel: GetMoviesRequestModel = {
            sortBy: mapSortOptionToRequestSortOption(sorting),
            sortOrder: SortOrder.Descending,
            search: capitalizeWords(genre),
            searchBy: SearchOptions.Genres,
        };

        if (query || genre === Genre.All) {
            requestModel.search = query;
            requestModel.searchBy = SearchOptions.Title;
        }

        filteringAndSortingAbortController.current = new AbortController();
        const requestParameters: RequestInit = {
            signal: filteringAndSortingAbortController.current!.signal,
        }

        getMovies(requestModel, requestParameters).then().then(movieModels => {
            setMovieList(movieModels);
            setMovieQuantity(movieModels.length)
        })
    }, [sorting, query, genre, searchParams, firstRender, filteringAndSortingAbortController, setSearchParams]);

    const movieContextModel: MovieContextModel = {
        selectedGenre: genre,
        setSelectedGenre: (genre: Genre) => {
            setSearchParams({ query: "", genre: genre, sorting: sorting })
        },
        moviesQuantity: moviesQuantity,
        movieList: movieList,
        searchQuery: query,
        setSearchQuery: (query: string) => {
            setSearchParams({ query: query, genre: Genre.All, sorting: sorting })
        },
        sortCriterion: sorting,
        setSortCriterion: (sorting: SortOption) => {
            setSearchParams({ query: query, genre: genre, sorting: sorting })
        },
    }

    return (
        <>
            <MovieContext.Provider value={movieContextModel}>
                <Outlet />
                <hr className="app__divider" />
                <MainPart />
                <Footer />
            </MovieContext.Provider>
        </>
    );
}
