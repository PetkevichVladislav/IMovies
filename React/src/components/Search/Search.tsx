import { useMovieContext } from '../MovieListPage/MovieListPage';
import './Search.scss';

import { useState, useCallback, useEffect, FC } from 'react';

interface ISearchProps { }

const Search: FC<ISearchProps> = () => {
    const movieContextModel = useMovieContext();
    const [isFocus, setIsFocus] = useState<boolean>(false);
    const [query, setQuery] = useState<string>(movieContextModel ? movieContextModel?.searchQuery : "");
    const internalSearch = useCallback(() => { movieContextModel?.setSearchQuery(query) }, [movieContextModel,query]);

    useEffect(() => {
        if (movieContextModel) {
            setQuery(movieContextModel.searchQuery);
        }
    }, [movieContextModel]);
    
    useEffect(() => {
        const handleKeyDown = (event: KeyboardEvent)  => {
            const isEnterPressed = event.key === "Enter";
            if (isEnterPressed && isFocus) {
                internalSearch();
            }
        };

        document.addEventListener("keydown", handleKeyDown);

        return () => {
            document.removeEventListener("keydown", handleKeyDown);
        };
    }, [isFocus, internalSearch])

    return (
        <div className="search">
            <input
                className="search__input"
                placeholder='What do you want to watch?'
                value={query}
                onChange={(e) => setQuery(e.target.value)}
                onFocus={() => setIsFocus(true)}
                onBlur={() => setIsFocus(false)}
            />
            <button className="search__button" onClick={internalSearch}>
                SEARCH
            </button>
        </div>
    )
}

export default Search;