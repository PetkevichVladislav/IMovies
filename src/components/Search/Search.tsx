import { MovieContextModel } from '../../models/context/MovieContextModel';
import { useMovieContext } from '../MovieListPage/MovieListPage';
import './Search.scss';

import { useState, useCallback, useEffect, FC } from 'react';

interface ISearchProps { }

const Search: FC<ISearchProps> = () => {
    const movieContextModel : MovieContextModel | null = useMovieContext();
    const [query, setQuery] = useState<string>(movieContextModel ? movieContextModel?.searchQuery : "");
    const internalSearch = useCallback(() => { movieContextModel?.setSearchQuery(query) }, [movieContextModel,query]);
    
    const [isFocus, setIsFocus] = useState<boolean>(false);
    useEffect(() => {
        const handleKeyDown = (event: KeyboardEvent)  => {
            const isEnterPressed = event.key === "Enter";
            if (isEnterPressed && isFocus) {
                internalSearch();
            }
        };

        document.addEventListener("keydown", handleKeyDown);

        return () => {
            document.removeEventListener("keydown", handleKeyDown);
        };
    }, [isFocus, internalSearch])

    return (
        <div className="search">
            <input
                className="search__input"
                placeholder='What do you want to watch?'
                value={query}
                onChange={(e) => setQuery(e.target.value)}
                onFocus={() => setIsFocus(true)}
                onBlur={() => setIsFocus(false)}
            />
            <button className="search__button" onClick={internalSearch}>
                SEARCH
            </button>
        </div>
    )
}

export default Search;