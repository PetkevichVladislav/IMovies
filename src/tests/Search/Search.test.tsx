import { fireEvent, render, screen } from '@testing-library/react';
import Search from '../../components/Search/Search';
import { MovieContext } from '../../components/MovieListPage/MovieListPage';
import { MovieContextModel } from '../../models/context/MovieContextModel';
import { SortOption } from '../../models/enum/SortOption';

describe('Search', () => {
    let context: MovieContextModel;

    beforeEach(() => {
        context = {
            movieList: [{ id: "", genres: [""], imagePath: "", overview: "", rating: 0, releaseDate: new Date(), runtime: 150, title: "" }],
            moviesQuantity: 0,
            searchQuery: "Action",
            selectedGenre: "",
            selectedMovieId: "",
            setSearchQuery: jest.fn(),
            setSelectedGenre: jest.fn(),
            setSelectedMovieId: jest.fn(),
            setSortCriterion: jest.fn(),
            sortCriterion: SortOption.ReleaseDate,
        }
    })

    it('renders correctly', () => {
        const { container } = render(
            <MovieContext.Provider value={context}>
                <Search />
            </MovieContext.Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it("When search creates, initial query should setups correctly", () => {
        render(
            <MovieContext.Provider value={context}>
                <Search />
            </MovieContext.Provider>
        );
        const input = screen.getByRole<HTMLInputElement>("textbox");
        expect(input.value).toBe(context.searchQuery);
    });


    it("When value in input changes query should be changed", () => {
        const changedValue = "Sci-Fi";
        context.setSearchQuery = jest.fn(() => context.searchQuery = changedValue);
        render(
            <MovieContext.Provider value={context}>
                <Search />
            </MovieContext.Provider>
        );
        const input = screen.getByRole<HTMLInputElement>("textbox");
        fireEvent.change(input, { target: { value: changedValue } });
        expect(input.value).toBe(changedValue);
    });

    it("When value in input changes and button clicks callback should return query", () => {
        const changedValue = "Sci-Fi";
        context.setSearchQuery = jest.fn(() => context.searchQuery = changedValue);
        render(
            <MovieContext.Provider value={context}>
                <Search />
            </MovieContext.Provider>
        );
        const input = screen.getByRole<HTMLInputElement>("textbox");
        fireEvent.change(input, { target: { value: changedValue } });
        const button = screen.getByRole("button", { name: "SEARCH" });
        fireEvent.click(button);
        expect(context.setSearchQuery).toHaveBeenCalledWith(changedValue);
    });

    it("When input is focussed and enter pressed callback should return query", () => {
        context.setSearchQuery = jest.fn();
        render(
            <MovieContext.Provider value={context}>
                <Search />
            </MovieContext.Provider>
        );
        const input = screen.getByRole<HTMLInputElement>("textbox");
        fireEvent.focus(input);
        fireEvent.keyDown(input, { key: 'Enter' });
        expect(context.setSearchQuery).toHaveBeenCalledWith(context.searchQuery);
    });

    it("When input is not focussed and enter pressed callback should return query", () => {
        context.setSearchQuery = jest.fn();
        render(
            <MovieContext.Provider value={context}>
                <Search />
            </MovieContext.Provider>
        );
        const input = screen.getByRole<HTMLInputElement>("textbox");
        fireEvent.keyDown(input, { key: 'Enter' });
        expect(context.setSearchQuery).not.toBeCalled();
    });
});