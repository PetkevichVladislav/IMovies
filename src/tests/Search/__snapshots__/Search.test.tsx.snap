// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search When data is valid component renders correctly 1`] = `
<div
  className="search"
>
  <input
    className="search__input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="What do you want to watch?"
    value="Action"
  />
  <button
    className="search__button"
    onClick={[Function]}
  >
    SEARCH
  </button>
</div>
`;
